index.js

```diff
@@ -196,14 +196,28 @@
     return _funcPath(false, path, implementations);
   }
 
+  /* istanbul ignore next */
+  //  functionName :: Function -> Nullable String
+  var functionName = 'name' in function f() {} ?
+      function functionName(f) { return f.name; } :
+      function functionName(f) {
+        var match = /function (\w*)/.exec(f);
+        return match && match[1];
+      };
+
   //  $ :: (String, Array TypeClass, StrMap (Array Location)) -> TypeClass
   function $(_name, dependencies, requirements) {
     function getBoundMethod(_name) {
       var name = 'fantasy-land/' + _name;
       return requirements[_name] === Constructor ?
         function(typeRep) {
-          return funcPath([name], typeRep) ||
-                 implPath([/function (\w*)/.exec(typeRep)[1], name]);
+          var f = funcPath([name], typeRep);
+          if (f == null && typeof typeRep === 'function') {
+            var typeIdent = functionName(typeRep);
+            /* istanbul ignore else */
+            if (typeIdent) f = implPath([typeIdent, name]);
```

Comment:
> The other option is to follow Irakli's suggestion by having functionName return '' rather than null in the failure case.

This is what I had in mind when I asked about the signature for `functionName`.

> Do we want the signature to be:
Function -> Nullable String or Function -> String?
I'm partial to the latter, but I know the code will cope with both.

The reason I would want to keep the check is because it will save us a loop in [`_funcPath`](https://github.com/gabejohnson/sanctuary-type-classes/blob/d2965b00f80d3c9dfb9d2da3906caa383ec6f2fb/index.js#L181)